src/interfaces/chain.cpp

```diff
@@ -0,0 +1,453 @@
+            CBlockIndex* block = stop_height ? ::chainActive[*stop_height] : ::chainActive.Tip();
+            while (block && block->nHeight >= start_height) {
+                if (!(block->nStatus & BLOCK_HAVE_DATA)) {
+                    return block->nHeight;
+                }
+                block = block->pprev;
+            }
+        }
+        return nullopt;
+    }
+    Optional<int> findFork(const uint256& hash, Optional<int>* height) override
+    {
+        const CBlockIndex *block{nullptr}, *fork{nullptr};
+        auto it = ::mapBlockIndex.find(hash);
+        if (it != ::mapBlockIndex.end()) {
+            block = it->second;
+            fork = ::chainActive.FindFork(block);
+        }
+        if (height) {
+            if (block) {
+                *height = block->nHeight;
+            } else {
+                height->reset();
+            }
+        }
+        if (fork) {
+            return fork->nHeight;
+        }
+        return nullopt;
+    }
+    bool isPotentialTip(const uint256& hash) override
+    {
+        if (::chainActive.Tip()->GetBlockHash() == hash) return true;
+        auto it = ::mapBlockIndex.find(hash);
+        return it != ::mapBlockIndex.end() && it->second->GetAncestor(::chainActive.Height()) == ::chainActive.Tip();
+    }
+    CBlockLocator getLocator() override { return ::chainActive.GetLocator(); }
+    Optional<int> findLocatorFork(const CBlockLocator& locator) override
+    {
+        if (CBlockIndex* fork = FindForkInGlobalIndex(::chainActive, locator)) {
+            return fork->nHeight;
+        }
+        return nullopt;
+    }
+    bool checkFinalTx(const CTransaction& tx) override
+    {
+        LockAnnotation lock(::cs_main);
+        return CheckFinalTx(tx);
+    }
+    bool acceptToMemoryPool(CTransactionRef tx, CValidationState& state) override
+    {
+        LockAnnotation lock(::cs_main);
+        return AcceptToMemoryPool(::mempool, state, tx, nullptr /* missing inputs */, nullptr /* txn replaced */,
+            false /* bypass limits */, ::maxTxFee /* absurd fee */);
+    }
+};
+
+class LockingStateImpl : public LockImpl, public UniqueLock<CCriticalSection>
+{
+    using UniqueLock::UniqueLock;
+};
+
+class HandlerImpl : public Handler
+{
+public:
+    HandlerImpl(Chain::Notifications& notifications) : m_forwarder(notifications) {}
+    ~HandlerImpl() override
+    {
+        // It is important to disconnect here in the HandlerImpl destructor,
+        // instead of in the HandleImpl::Forwarder destructor, because
+        // UnregisterAllValidationInterfaces() internally accesses Forwarder
+        // virtual methods which are no longer available after Forwarder object
+        // has begun to be being destroyed.
+        m_forwarder.disconnect();
+    }
+    void disconnect() override { m_forwarder.disconnect(); }
+
+    struct Forwarder : CValidationInterface
+    {
+        Forwarder(Chain::Notifications& notifications) : m_notifications(&notifications)
```

Comment:
```
2018-09-19 14:39:06 clang-tidy(pr=10973): src/interfaces/chain.cpp:182:9: warning: single-argument constructors must be marked explicit to avoid unintentional implicit conversions [google-explicit-constructor]
```